rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,215 @@
+
+#### Build ZoneRecord from the provider zone
+
+When the DNS Operator has finished reconciling a deleting kuadrantRecord it will pull the most recent records from the DNS Provider and construct a separate zoneRecord for the same root host as the deleting kuadrantRecord.
+
+#### Prune the zoneRecord
+
+It will recursively analyse the zoneRecord and ensure that it is structurally sound and that all records can be resolved to at least one external CNAME or A record.
+
+If during this pruning period the zoneRecord is modified, then these modifications will need to be sent to the DNS Provider to clean the zone.
+
+Once completed, the zoneRecord can be deleted (this may be possible to only reflect in memory completely).
+
+#### Removing the finalizer from kuadrantRecords
+
+When a kuadrantRecord is reconciled, the DNS Operator should apply a finalizer to it.
+
+When a kuadrantRecord is being removed, the following must be successfully completed before the finalizer can be removed:
+- Remove the local clusters records and targets from the relevant zone
+- Perform a prune on the relevant root host
+- Apply the results of the prune to the DNS Provider
+
+Only once these actions have all resolved should the finalizer be removed, and the kuadrantRecord allowed to be deleted.
+
+Should a deleted DNS Record be recreated, or a deleted DNS Policy be restored, all of the deleted records will be restored by the DNS Operator, at that point.
+
+#### Aggregating Status
+
+When the DNS Operator performs actions based on the state of a kuadrantRecord, it should update the kuadrantRecord status with results of these actions, similar to how it is currently implemented.
+
+When the Kuadrant operator sees these statuses on the kuadrantRecords related to a DNS Policy, it should aggregate them into a consolidated status on the DNS Policy status.
+
+It will be possible from this DNS Policy status to determine that all the records for this cluster, related to this DNS Policy, have been accepted by the DNS Provider.
+
+#### Routing Strategies
+
+There are currently 2 routing strategies "simple" and "loadbalanced".
+
+##### Simple
+The simple strategy is also compatible with ExternalDNS which makes it useful in some use-cases. This strategy outputs a very simplistic zone file and will only be functional with a single cluster.
+
+##### Loadbalanced
+This strategy outputs a DNS record tree with various levels of CNAMEs to allow for geo loadbalancing as well as loadbalancing within a geo and is compatible with multiple clusters.
+
+This can be seen here: ![Load Balanced DNS Records](0008-distributed-dns-assets%2F0008-loadbalanced-dns-records.jpg)
+
+##### Migration
+
+These 2 strategies are not compatible, and as such the RoutingStrategy field will be set to immutable, requiring the deletion and recreation of the DNS Policy in order to update this value. This will result in the related records being removed from the zone before being created in the new format.
```

Comment:
THis will result in a flip/flop scenerio and unresolvable state which will be spotted by the max back off validation check mentioned above and flagged in the status

Reply:
RFC wil be updated to make this clearer